File: daytona-terminal/backend/app/api/sessions.py
"""Terminal session API routes."""

from fastapi import APIRouter, HTTPException, status

from app.models.session import (
    TerminalResize,
    TerminalSession,
    TerminalSessionCreate,
)
from app.services.terminal import TerminalError, get_terminal_service

router = APIRouter(prefix="/sessions", tags=["sessions"])


@router.post("", response_model=TerminalSession, status_code=status.HTTP_201_CREATED)
async def create_session(request: TerminalSessionCreate):
    """Create a new terminal session."""
    try:
        service = get_terminal_service()
        session = await service.create_session(request)
        return session
    except TerminalError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )


@router.get("", response_model=list[TerminalSession])
async def list_sessions(workspace_id: str | None = None):
    """List all terminal sessions."""
    service = get_terminal_service()
    sessions = await service.list_sessions(workspace_id)
    return sessions


@router.get("/{session_id}", response_model=TerminalSession)
async def get_session(session_id: str):
    """Get a session by ID."""
    service = get_terminal_service()
    session = await service.get_session(session_id)
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Session not found: {session_id}",
        )
    return session


@router.delete("/{session_id}", status_code=status.HTTP_204_NO_CONTENT)
async def close_session(session_id: str):
    """Close a terminal session."""
    service = get_terminal_service()
    success = await service.close_session(session_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Session not found: {session_id}",
        )


@router.post("/{session_id}/resize", response_model=TerminalSession)
async def resize_session(session_id: str, request: TerminalResize):
    """Resize a terminal session."""
    service = get_terminal_service()
    try:
        await service.resize_terminal(session_id, request.cols, request.rows)
        session = await service.get_session(session_id)
        if not session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Session not found: {session_id}",
            )
        return session
    except TerminalError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )


@router.get("/{session_id}/context")
async def get_terminal_context(session_id: str, max_bytes: int = 10000):
    """Get recent terminal output for AI context."""
    service = get_terminal_service()
    session = await service.get_session(session_id)
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Session not found: {session_id}",
        )
    
    context = service.get_terminal_context(session_id, max_bytes)
    return {"context": context}

File: daytona-terminal/backend/src/api/sessions.ts
import type { FastifyPluginAsync } from 'fastify';
import { z } from 'zod';
import { terminalService } from '../services/terminal.service.js';
import { TerminalSessionCreateSchema, TerminalResizeSchema } from '../models/session.js';

export const sessionsRoutes: FastifyPluginAsync = async (fastify) => {
  // List sessions
  fastify.get('/', async (request) => {
    const { workspaceId } = request.query as { workspaceId?: string };
    return terminalService.listSessions(workspaceId);
  });

  // Create session
  fastify.post('/', async (request, reply) => {
    try {
      const body = TerminalSessionCreateSchema.parse(request.body);
      const session = await terminalService.createSession(body);
      reply.code(201);
      return session;
    } catch (error) {
      if (error instanceof z.ZodError) {
        reply.code(400);
        return { error: 'Validation error', details: error.errors };
      }
      if (error instanceof Error) {
        reply.code(400);
        return { error: error.message };
      }
      throw error;
    }
  });

  // Get session
  fastify.get('/:id', async (request, reply) => {
    const { id } = request.params as { id: string };
    const session = await terminalService.getSession(id);
    if (!session) {
      reply.code(404);
      return { error: `Session not found: ${id}` };
    }
    return session;
  });

  // Close session
  fastify.delete('/:id', async (request, reply) => {
    const { id } = request.params as { id: string };
    const success = await terminalService.closeSession(id);
    if (!success) {
      reply.code(404);
      return { error: `Session not found: ${id}` };
    }
    reply.code(204);
    return;
  });

  // Resize session
  fastify.post('/:id/resize', async (request, reply) => {
    const { id } = request.params as { id: string };
    try {
      const { cols, rows } = TerminalResizeSchema.parse(request.body);
      await terminalService.resize(id, cols, rows);
      const session = await terminalService.getSession(id);
      return session;
    } catch (error) {
      if (error instanceof z.ZodError) {
        reply.code(400);
        return { error: 'Validation error', details: error.errors };
      }
      if (error instanceof Error) {
        reply.code(400);
        return { error: error.message };
      }
      throw error;
    }
  });

  // Get terminal context (for AI)
  fastify.get('/:id/context', async (request, reply) => {
    const { id } = request.params as { id: string };
    const { maxBytes = 10000 } = request.query as { maxBytes?: number };

    const session = await terminalService.getSession(id);
    if (!session) {
      reply.code(404);
      return { error: `Session not found: ${id}` };
    }

    const context = terminalService.getContext(id, maxBytes);
    return { context };
  });

  // Get command blocks
  fastify.get('/:id/blocks', async (request, reply) => {
    const { id } = request.params as { id: string };
    const { limit = 20 } = request.query as { limit?: number };

    const session = await terminalService.getSession(id);
    if (!session) {
      reply.code(404);
      return { error: `Session not found: ${id}` };
    }

    const blocks = terminalService.getBlocks(id, limit);
    return { blocks };
  });

  // Get last block (for quick error checking)
  fastify.get('/:id/blocks/last', async (request, reply) => {
    const { id } = request.params as { id: string };

    const session = await terminalService.getSession(id);
    if (!session) {
      reply.code(404);
      return { error: `Session not found: ${id}` };
    }

    const block = terminalService.getLastBlock(id);
    return { block };
  });
};
